compiler/rustc_data_structures/src/base_n.rs

```diff
@@ -42,3 +43,52 @@ pub fn encode(n: u128, base: usize) -> String {
     push_str(n, base, &mut s);
     s
 }
+
+// This trait just lets us reserve the exact right amount of space when doing fixed-length
+// case-insensitve encoding. Add any impls you need.
+pub trait Base36Encodable: Copy + Into<u128> {
+    fn encoded_len() -> usize;
+}
+
+impl Base36Encodable for u128 {
+    fn encoded_len() -> usize {
+        25
+    }
+}
+
+impl Base36Encodable for u64 {
+    fn encoded_len() -> usize {
+        13
+    }
+}
+
+impl Base36Encodable for u32 {
+    fn encoded_len() -> usize {
+        7
+    }
+}
+
+pub fn push_case_insensitive<N: Base36Encodable>(n: N, output: &mut String) {
+    // SAFETY: We will only append ASCII bytes.
```

Comment:
I didn't do that because I don't know what to do with this diagnostic:
```
error: unconstrained generic constant
  --> compiler/rustc_data_structures/src/base_n.rs:76:25
   |
76 |     let mut buf = [0u8; N::LEN];
   |                         ^^^^^^
   |
   = help: try adding a `where` bound using this expression: `where [(); N::LEN]:`
```

Reply:
Being able to use a stack buffer here would be very elegant because then I could justify to myself the design of writing this as a Display helper that does a single `write_str` call, and that would integrate very nicely with how this formatting is used.

Reply:
Hm. I'll try that.